Cypress: Initial implementation of MCUBoot for 062 - BSP less

romanjoe opened the pull request:
This PR is created to substitute existing one https://github.com/JuulLabs-OSS/mcuboot/pull/605.

This implementation is not dependent on BSP asset, as previous one. All the dependencies used as submodules are publicly available.

Features:

- All the code is contained in boot/cypress folder and do not change any MCUBoot source code
- Makefile based build support implemented
- Dependencies are provided as submodules
- image signature and hash validation implemented using software mbedtls
- BlinkyApp - test application provided to quickly test MCUBootApp which is a bootloader.
- Information regarding bootloader and test application build can be found in corresponding Readme.md files.

utzig commented:
Please rebase on `master`.

utzig commented:
@romanjoe This still adds a `mbedtls` submodule from cypress, can't we use the one we already have at `ext/mbedtls`?

romanjoe commented:
> @romanjoe This still adds a `mbedtls` submodule from cypress, can't we use the one we already have at `ext/mbedtls`?

Looks like it is the same instance, as we have - from base https://github.com/ARMmbed/mbedtls, so i suppose we can substitute. When we first started this work, there were some subset of mbedtls in upstream, that is why we added full repo to our folders.

utzig commented on boot/cypress/MCUBootApp/image_ec256_mbedtls.c:
```diff
@@ -0,0 +1,166 @@
+/*
```
This file could probably be refactored to go into `boot/bootutil/src/image_ec256.c`.

romanjoe commented:
- fixed code per Fabio comments
- rebased this branch against Fabios PR with MbedTLS version update
- got rid of cypress instance of MbedTLS, using ext/mbedtls now
- updated makefiles accordingly

utzig commented on boot/cypress/BlinkyApp/BlinkyApp.mk:
```diff
@@ -14,7 +14,7 @@
 # You may obtain a copy of the License at
 #
-#     http://www.apache.org/licenses/LICENSE-2.0
+#	http://www.apache.org/licenses/LICENSE-2.0
```
This should be a space

utzig commented on boot/cypress/BlinkyApp/BlinkyApp.mk:
```diff
@@ -25,9 +25,9 @@
-#   - image type to BOOT
+#	- compiler GCC
+#	- build configuration to Debug
+#	- image type to BOOT
```
These would align better if they were spaces

utzig commented on boot/cypress/BlinkyApp/libs.mk:
```diff
@@ -14,7 +14,7 @@
 # You may obtain a copy of the License at
 #
-#     http://www.apache.org/licenses/LICENSE-2.0
+#	http://www.apache.org/licenses/LICENSE-2.0
```
tab -> space

utzig commented on boot/cypress/BlinkyApp/platforms.mk:
```diff
@@ -14,7 +14,7 @@
 # You may obtain a copy of the License at
 #
-#     http://www.apache.org/licenses/LICENSE-2.0
+#	http://www.apache.org/licenses/LICENSE-2.0
```
tab -> space

utzig commented on boot/cypress/BlinkyApp/toolchains.mk:
```diff
@@ -14,7 +14,7 @@
 # You may obtain a copy of the License at
 #
-#     http://www.apache.org/licenses/LICENSE-2.0
+#	http://www.apache.org/licenses/LICENSE-2.0
```
tab -> space

utzig commented on boot/cypress/BlinkyApp/toolchains.mk:
```diff
@@ -24,10 +24,10 @@
+GCC_ARM	:= 1
+IAR		:= 2
+ARM		:= 3
+OTHER 	:= 4
```
Would align better if was using spaces here

d3zd3z commented:
I think this looks ok, as most of the change is within the Cypress directory. I'll be curious if the extra submodules end up being challenging for people on other targets.

I still think we should try to get Cypress support through the various RTOS's. Effectively, this is adding another RTOS, that just happens to be a lot thinner. But, this isn't really going to scale for adding more targets this way.

utzig commented:
Could you please squash f0e95a02e7d8e50f65c41d382cb97ebb9bd82f2a, 0a4a53c9c9de9d119ae8f9ea38a4a5739919f507 and 8a44b8d275702e5e64e9e41bebd13bc072288525?
